src/logic/graphFrames.ts

```diff
@@ -1,19 +1,95 @@
-import { Frame, DissasembledGraph, Algorithm } from './commonInterfaces';
+import {
+  Frame,
+  DisassembledGraph,
+  Algorithm,
+  Edge,
+  TraversalRecord,
+  Paths,
+  GraphIndex,
+} from './commonInterfaces';
+
+interface FrameWithEdges {
+  activeEdges: Array<Edge>,
+  inactiveEdges: Array<Edge>,
+};
 
 export default function createFrames(
-  dissasembledGraph: DissasembledGraph,
+  dissasembledGraph: DisassembledGraph,
   algorithm: Algorithm,
 ):
   Array<Frame>
 {
-  return [ // Temporary fixed result
-    { //frame 1
-      inactiveEdges: [],
-      activeEdges: [dissasembledGraph[1][0]],
-    },
-    { //frame 2
-      inactiveEdges: [dissasembledGraph[1][0]],
-      activeEdges: [],
-    },
-  ]
+  const edgeFrames = createFramesWithEdges(algorithm);
+  return convertToIndexes(dissasembledGraph, edgeFrames);
+}
+
+export function createFramesWithEdges(from: Algorithm) {
+  let frames: Array<FrameWithEdges> = [];
+  for (let tr of from.traversalHistory) {
+    frames.push(createFrameForNode(frames));
+    frames.push(...createFramesForLookups(tr, frames));
+  }
+  return frames;
+}
+
+function createFrameForNode(frames: Array<FrameWithEdges>): FrameWithEdges {
+  const lastFrame = frames[frames.length - 1] || { activeEdges: [], inactiveEdges: [] };
+  return {
+    activeEdges: [],
+    inactiveEdges: lastFrame.activeEdges.concat(lastFrame.inactiveEdges),
+  };
+}
+
+function createFramesForLookups(
+  tr: TraversalRecord,
+  frames: Array<FrameWithEdges>,
+):
+  Array<FrameWithEdges>
+{
+  const nodeFrames: Array<FrameWithEdges> = [
+    frames[frames.length - 1] || { activeEdges: [], inactiveEdges: [] },
+  ];
+  for (let toNode of tr.lookups) {
+    const lastFrame = nodeFrames[nodeFrames.length - 1];
+    nodeFrames.push({
+      activeEdges: [new Edge(tr.node, toNode)],
+      inactiveEdges: lastFrame.activeEdges.concat(lastFrame.inactiveEdges),
+    });
+  }
+  nodeFrames.shift() //First frame was repeated
+  return nodeFrames;
+}
+
+function convertToIndexes(
+  graph: DisassembledGraph,
+  edgeFrames: Array<FrameWithEdges>,
+):
+  Array<Frame>
+{
+  const frames: Array<Frame> = [];
+  const graphEdges = graph[GraphIndex.EDGES];
+  for (let eFrame of edgeFrames) {
+    frames.push({
+      activeEdges: convertEdgeArrayToIdexes(eFrame.activeEdges, graphEdges),
+      inactiveEdges: convertEdgeArrayToIdexes(eFrame.inactiveEdges, graphEdges),
+    })
+  }
+  return frames
+}
+
+function convertEdgeArrayToIdexes(
+  edgeArray: Array<Edge>,
+  graphEdges: Array<Edge>,
+):
+  Array<number>
+{
+  const indexes: Set<number> = new Set();
+  for (let edge of edgeArray) {
```

Comment:
The more for loops are in the code the more obscure the logic gets, I'm not a JS expert but I think you can use [Array.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) here. (Ideally, convert previous for loops whereas possible too)

Reply:
It needs to be a set. `Array.foreach` would be good pick. I don't really know advanced methods for sets. Probably there is some weird `Set.from(...)` syntax that would make it even better

Reply:
Also creating and modifing something instantly in next line is a bit of antipattern, but I'll leave it with `foreach`.

Reply:
I think you can simply pass the array
https://stackoverflow.com/questions/28965112/javascript-array-to-set